lib/src/rules/avoid_operator_equals_on_mutable_classes.dart

```diff
@@ -0,0 +1,103 @@
+// Copyright (c) 2019, the Dart project authors. Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'package:analyzer/dart/ast/ast.dart';
+import 'package:analyzer/dart/ast/token.dart';
+import 'package:analyzer/dart/ast/visitor.dart';
+import 'package:analyzer/dart/element/element.dart';
+import 'package:linter/src/analyzer.dart';
+import 'package:linter/src/ast.dart';
+
+const _desc =
+    r'AVOID overloading operator == on classes not marked `@immutable`.';
+
+const _details = r'''
+
+**AVOID** overloading operator == on classes not marked `@immutable`.
+
+If a class is not immutable, overloading operator == and hashCode can lead to
+unpredictable and undesirable behavior when used in collections. See
+https://dart.dev/guides/language/effective-dart/design#avoid-defining-custom-equality-for-mutable-classes
+for more information.
+
+**GOOD:**
+```
+@immutable
+class A {
+  final String key;
+  const A(this.key);
+  @override operator ==(other) => other is A && other.key == key;
+  @override int hashCode() => key.hashCode;
+}
+```
+
+**BAD:**
+```
+class B {
+  final String key;
+  const B(this.key);
+  @override operator ==(other) => other is B && other.key == key;
+  @override int hashCode() => key.hashCode;
+}
+```
+
+''';
```

Comment:
As someone relatively new to dart this suggests to me that the error is really that `B` could be marked `@immutable` but isn't.

Reply:
I was trying to emphasize that the lint is literally checking for the `@immutable` annotation rather than doing an analysis of the mutability of the class: here, the lint will trigger even though the class is not mutable. If you think it's clearer I can remove the `final` keyword from the field, or include both examples.

Reply:
I added another example and some explanatory text.